src/iniconfig/__init__.py

```diff
@@ -1,164 +1,175 @@
 """ brain-dead simple parser for ini-style files.
 (C) Ronny Pfannschmidt, Holger Krekel -- MIT licensed
 """
-__all__ = ["IniConfig", "ParseError"]
+from __future__ import annotations
+from typing import (
+    Callable,
+    Iterator,
+    Mapping,
+    Optional,
+    Tuple,
+    TypeVar,
+    Union,
+    TYPE_CHECKING,
+    NoReturn,
+    NamedTuple,
+    overload,
+    cast,
+)
 
-COMMENTCHARS = "#;"
 
+if TYPE_CHECKING:
+    from typing_extensions import Final
 
-class ParseError(Exception):
-    def __init__(self, path, lineno, msg):
-        Exception.__init__(self, path, lineno, msg)
-        self.path = path
-        self.lineno = lineno
-        self.msg = msg
+__all__ = ["IniConfig", "ParseError", "COMMENTCHARS", "iscommentline"]
 
-    def __str__(self):
-        return f"{self.path}:{self.lineno + 1}: {self.msg}"
+from .exceptions import ParseError
+from . import _parse
+from ._parse import COMMENTCHARS, iscommentline
+
+_D = TypeVar("_D")
+_T = TypeVar("_T")
 
 
 class SectionWrapper:
-    def __init__(self, config, name):
+    config: Final[IniConfig]
+    name: Final[str]
+
+    def __init__(self, config: IniConfig, name: str):
         self.config = config
         self.name = name
 
-    def lineof(self, name):
+    def lineof(self, name: str) -> int | None:
         return self.config.lineof(self.name, name)
 
-    def get(self, key, default=None, convert=str):
+    def get(
+        self,
+        key: str,
+        default: _D | None = None,
+        convert: Callable[[str], _T] | None = None,
+    ) -> _D | _T | str | None:
         return self.config.get(self.name, key, convert=convert, default=default)
 
-    def __getitem__(self, key):
+    def __getitem__(self, key: str) -> str:
         return self.config.sections[self.name][key]
 
-    def __iter__(self):
-        section = self.config.sections.get(self.name, [])
+    def __iter__(self) -> Iterator[str]:
+        section: Mapping[str, str] = self.config.sections.get(self.name, {})
 
-        def lineof(key):
-            return self.config.lineof(self.name, key)
+        def lineof(key: str) -> int:
+            return self.config.lineof(self.name, key)  # type: ignore
```

Comment:
I recommend always adding the error code to `type: ignores`, e.g. `type: ignore[arg-type]` (just an example, I'm not sure what is the error code in this case)